src/System.Private.CoreLib/shared/System/Text/Rune.cs

```diff
@@ -210,6 +211,457 @@ private static Rune ChangeCaseCultureAware(Rune rune, TextInfo textInfo, bool to
+                    goto InvalidData;
+                }
+            }
+
+            // If we got to this point, the input buffer was empty, or the buffer
+            // was a single element in length and that element was a high surrogate char.
+
+            charsConsumed = source.Length;
+            result = ReplacementChar;
+            return OperationStatus.NeedMoreData;
+
+        InvalidData:
+
+            charsConsumed = 1; // maximal invalid subsequence for UTF-16 is always a single code unit in length
+            result = ReplacementChar;
+            return OperationStatus.InvalidData;
+        }
+
+        /// <summary>
+        /// Decodes the <see cref="Rune"/> at the end of the provided UTF-16 source buffer.
+        /// </summary>
+        /// <remarks>
+        /// This method is very similar to <see cref="Decode(ReadOnlySpan{char}, out Rune, out int)"/>, but it allows
+        /// the caller to loop backward instead of forward. The typical calling convention is that on each iteration
+        /// of the loop, the caller should slice off the final <paramref name="charsConsumed"/> elements of
+        /// the <paramref name="source"/> buffer.
+        /// </remarks>
+        public static OperationStatus DecodeFromEnd(ReadOnlySpan<char> source, out Rune result, out int charsConsumed)
+        {
+            int index = source.Length - 1;
+            if ((uint)index < (uint)source.Length)
+            {
+                // First, check for the common case of a BMP scalar value.
+                // If this is correct, return immediately.
+
+                char finalChar = source[index];
+                if (TryCreate(finalChar, out result))
+                {
+                    charsConsumed = 1;
+                    return OperationStatus.Done;
+                }
+
+                if (char.IsLowSurrogate(finalChar))
+                {
+                    // The final character was a UTF-16 low surrogate code point.
+                    // This must be preceded by a UTF-16 high surrogate code point, otherwise
+                    // we have a standalone low surrogate, which is always invalid.
+
+                    index--;
+                    if ((uint)index < (uint)source.Length)
+                    {
+                        char penultimateChar = source[index];
+                        if (TryCreate(penultimateChar, finalChar, out result))
+                        {
+                            // Success! Formed a supplementary scalar value.
+                            charsConsumed = 2;
+                            return OperationStatus.Done;
+                        }
+                    }
+
+                    // If we got to this point, we saw a standalone low surrogate
+                    // and must report an error.
+
+                    charsConsumed = 1; // standalone surrogate
+                    result = ReplacementChar;
+                    return OperationStatus.InvalidData;
```

Comment:
why we don't return NeedMoreData when having low surrogate only? we don't have the case not providing the whole needed length of the span even from the end?

Reply:
The general idea is that this invariant should hold:

```cs
// Do a reverse read
OperationStatus reverseStatus = Rune.DecodeFromEnd(input, out Rune reverseValue, out int reverseCharsRead);

// Slice the input so that it contains _only_ what we just read
input = input.Slice(input.Length - reverseCharsRead);

// Now do a forward read and make sure all of the return values are the same
OperationStatus forwardStatus = Rune.Decode(input, out Rune forwardValue, out int forwardCharsRead);

Assert.Equal(reverseStatus, forwardStatus);
Assert.Equal(reverseValue, forwardValue);
Assert.Equal(reverseCharsRead, forwardCharsRead);
```

In other words, imagine doing a forward read in a loop until one read returns "I've consumed everything that remains in the buffer." (This is detected via `charsConsumed == input.Length`.) This final read tells you what the last `Rune` in the input was or gives you information about why the sequence is invalid or incomplete. A reverse read should return this exact same value.

This also implies that a standalone low surrogate is invalid, not incomplete, since we assume that there's never any extra data _before_ the input. The only time a standalone surrogate can ever be anything other than invalid is when it's at the end of the buffer, and only if it's a high surrogate, at which point we return "This _might_ be valid, but I need more data to make a proper determination."